File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: Option<Box<ListNode>>,
// }
//
// impl ListNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         ListNode { next: None, val }
//     }
// }

impl Solution {
    pub fn remove_nth_from_end(head: Option<Box<ListNode>>, n: i32) -> Option<Box<ListNode>> {
        let mut cursor = head.as_ref();
        let mut length = 0;
        while let Some(node) = cursor{
          length += 1;
          cursor = node.next.as_ref();
        }

        let target = length - n;


        let mut index = 0;
        let mut tmp = Box::new(ListNode::new(0));
        tmp.next = head;
        // println!("{:?}",tmp);
        let mut dummy = Some(tmp);
        let mut previous = dummy.as_mut();
        // cursor = dummy.as_ref().unwrap().next.as_ref();

        // println!("{:?}",previous);
        // println!("{:?}",cursor);
      

        while let Some(node) = previous{ 
          // println!("loop {}",index);
          // println!("{:?}",cursor);
          println!("{:?}",node);
          // println!("{:?}",node.next);
          // println!("{:?}",node.next.as_ref().unwrap().next);
          if index == target{
            println!("hello!");
            node.next = node.next.as_ref().unwrap().next.clone();
            // println!("{:?}",previous.unwrap().next);
            // let next_value = node.next.as_ref();
            // previous.unwrap().next ;
            // previous.unwrap().next = Some(node.next.unwrap());
          }
          previous = node.next.as_mut();
          index += 1;
        }

        return dummy.unwrap().next;

    }
}

File: Data Structures & Algorithms/add-two-numbers/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        
        int carry = 0;
        ListNode dummy = new ListNode(0);
        ListNode result = dummy;
        while (l1 != null || l2 != null || carry != 0){
            int value1 = 0;
            int value2 = 0;

            if (l1 != null){
                value1 = l1.val;
            }
            if (l2 != null){
                value2 = l2.val;
            }

            int sum = value1 + value2 + carry;
            if (sum >= 10){
                sum = sum - 10;
                carry = 1;
            }
            else{
                carry = 0;
            }

            result.next = new ListNode(sum);
            result = result.next;
            if (l1 != null){
                l1 = l1.next;
            }
            if (l2 != null){
                l2 = l2.next;
            }
        }

        return dummy.next;

    }
}
